### PR TITLE
chore: uses new subscriptions table/drawer layout for zoneIngress

### PR DESCRIPTION
**Pull request by johncowen**
Adds new Subscription Table/Drawer UI to ZoneIngresses (whilst maintaining old-style for Egress and Dataplanes until we change those also)

Part of: https://github.com/kumahq/kuma-gui/issues/2921

---

~Lots of failing tests here, so keeping in draft until I've sorted that. I would guess it's probably chore-like updating of tests but we'll see!~

Turns out we can't use `addRouteName` from https://github.com/kumahq/kuma-gui/pull/3151 due to things being slightly different and weird between dev source and compiled/built source.

Instead I reverted to use route config `props` here, which was second favourite option. I plan to improve on this a little more, and if I ever figure out a more reliable way to add a dynamic global like `$routeName` I can add it at a later date. In the meantime I'd rather keep this table/drawers feature on track. Apologies for the noisy PR!

### CODE REVIEW

File: src/app/application/components/route-view/RouteView.vue [hunk 1]
```diff
@@ -198,6 +198,9 @@ watch(() => {
 }, { immediate: true })
 
 watch(() => props.name, () => {
+  if (props.name.startsWith('$routeName')) {
+    throw new Error('`$routeName` has been used for a name, please use `:name` not `name`')
+  }
```
Comment (johncowen) [inline src/app/application/components/route-view/RouteView.vue:203]: I bumped into a thing were I used `name="$routeName!"` instead of `:name="$routeName!"`, and it took me a good few seconds/minutes to realize.

This should prevent that from happening again, and seeing as we are already watching `props.name` it was no hassle to add.

File: src/app/kuma/locales/en-us/index.yaml [hunk 2]
```diff
@@ -94,10 +94,13 @@ http:
       tls: TLS
       mtls: mTLS
       mTLS: mTLS
+      id: ID
       globalInstanceId: Global instance ID
       controlPlaneInstanceId: CP instance ID
+      zoneInstanceId: Zone Leader Instance ID
       connectTime: Connected
       disconnectTime: Disconnected
+      version: Version
```
Comment (johncowen) [inline src/app/kuma/locales/en-us/index.yaml:103]: These are now part of the `http.api.property` i18n string. Means 2 things:

1. We don't have to add them in other modules as these are commonly used across the entire app.
2. The terms are consistent through out the entire app (see point 1)

This is also related to https://github.com/kumahq/kuma-gui/issues/2814 and https://github.com/kumahq/kuma-gui/issues/2431.

You'll notice I possibly seem overly cautious about adding i18n at the moment even though we really want our entire app to be 100% i18n'ed. I'd just really like to find/figure a good pattern/tooling so we don't just keep digging the current hole deeper. The hole is currently on the wrong place, and once we find the right place, we are gonna have to fill this one back in again. Possibly bad analogy! 😆

File: src/app/subscriptions/components/SubscriptionList.vue [hunk 3]
```diff
@@ -28,12 +28,12 @@ import { computed } from 'vue'
...
 import SubscriptionDetails from './SubscriptionDetails.vue'
 import SubscriptionHeader from './SubscriptionHeader.vue'
+import type { Subscription } from '../data'
 import AccordionItem from '@/app/common/AccordionItem.vue'
 import AccordionList from '@/app/common/AccordionList.vue'
-import type { DiscoverySubscription, KDSSubscription } from '@/types/index.d'
 
 const props = defineProps<{
-  subscriptions: Array<KDSSubscription | DiscoverySubscription>
+  subscriptions: Subscription[]
```
Comment (johncowen) [inline src/app/subscriptions/components/SubscriptionList.vue:36]: I had to amend the current Subscription{Header, List, Details} components in order to keep supporting subscriptions for Dataplanes seeing as I haven't upgrade that yet (thats coming next)

File: src/app/subscriptions/data/index.ts [hunk 4]
```diff
@@ -49,3 +94,4 @@ export const SubscriptionCollection = {
     }
   },
 }
+export type SubscriptionCollection = ReturnType<typeof SubscriptionCollection.fromArray>
```
Comment (johncowen) [inline src/app/subscriptions/data/index.ts:97]: Changes here to our transformation layer, partly an upgrade to the style here, but also brings parts of KDS and DiscoverySubscriptions together so we can just refer to a `Subscription` instead of a union of the two.

File: src/app/subscriptions/routes.ts [hunk 5]
```diff
@@ -1,20 +1,20 @@
...
         {
           path: 'overview',
-          name: 'subscription-summary-overview-view',
+          name: `${prefix}-subscription-summary-overview-view`,
           component: () => import('@/app/subscriptions/views/SubscriptionSummaryOverviewView.vue'),
         },
         {
           path: 'config',
-          name: 'subscription-summary-config-view',
+          name: `${prefix}-subscription-summary-config-view`,
```
Comment (johncowen) [inline src/app/subscriptions/routes.ts:17]: Changes relating to "mounting" a modules routes at different "mount points".

File: src/app/subscriptions/views/SubscriptionSummaryConfigView.vue [hunk 6]
```diff
@@ -6,36 +6,28 @@
...
+        language="yaml"
+        :code="YAML.stringify(props.data.$raw)"
+        is-searchable
+        :query="route.params.codeSearch"
+        :is-filter-mode="route.params.codeFilter"
+        :is-reg-exp-mode="route.params.codeRegExp"
+        @query-change="route.update({ codeSearch: $event })"
+        @filter-mode-change="route.update({ codeFilter: $event })"
+        @reg-exp-mode-change="route.update({ codeRegExp: $event })"
+      />
```
Comment (johncowen) [inline src/app/subscriptions/views/SubscriptionSummaryConfigView.vue:22]: I noticed the SubscriptionConfig tab content had a header and a KCard which is not how we render this in other areas of the GUI so I removed those things

File: src/app/zones/data/index.ts [hunk 7]
```diff
@@ -114,3 +103,4 @@ export const ZoneOverview = {
     }
   },
 }
+export type ZoneOverview = ReturnType<typeof ZoneOverview.fromObject>
```
Comment (johncowen) [inline src/app/zones/data/index.ts:106]: Upgraded this to use the slightly different new inference based approach to our data transformers

File: src/test-support/mocks/src/zone-ingresses/_/_overview.ts [hunk 8]
```diff
@@ -57,27 +57,41 @@ export default ({ fake, env }: EndpointDependencies): MockResponder => (req) =>
...
+                            responsesAcknowledged: Number,
+                            responsesRejected: Number,
+                          }, value,
+                        )
+                        return prev
+                      },
+                      {} as Record<string, {}>,
+                    ),
+                  }
+                })(),
```
Comment (johncowen) [inline src/test-support/mocks/src/zone-ingresses/_/_overview.ts:94]: Much better realistic randomness for the ack/nacks mocks. This will eventually looks slightly better if we add deep functionality to `partitionInto` (see TODO comment I left a little further up)

File: src/types/index.d.ts [hunk 9]
```diff
@@ -238,33 +215,59 @@ export interface EnvoyVersion {
...
-  id: string
   controlPlaneInstanceId: string
-  connectTime?: string
-  disconnectTime?: string
-  status: DiscoverySubscriptionStatus
   generation?: number
-  version?: Version
-}
+  status: DiscoverySubscriptionStatus
+} & Subscription
```
Comment (johncowen) [inline src/types/index.d.ts:270]: These are the old handwritten types that we still need to support until we get these types OpenAPI'ed.

I found these to be slightly wrong, and I also moved them together in the file so I could look at the in the same place without a split. Whenever you are working on KDSSubscriptions you invariably need to work on DiscoverySubscription , much better if they are next to each other.

**Review by johncowen — COMMENTED** [review]
This PR ended up needing to be a bit bigger than I'd wanted, so I left some inline notes also:

File: src/app/application/components/route-view/RouteView.vue (hunk 1 above)
Comment (johncowen) [reply src/app/application/components/route-view/RouteView.vue]: Following my discovery that we can't use a global `$routeName` I remove this check

File: src/app/kuma/locales/en-us/index.yaml (hunk 2 above)
Comment (schogges) [reply src/app/kuma/locales/en-us/index.yaml:103]: I like the idea to have the entire app i18n'ed 👍 but you are right that we have to find a good pattern and structure 🙂 what do you mean with "the hole is currently on the wrong place"?

File: src/app/zone-ingresses/views/ZoneIngressDetailView.vue [hunk 10]
```diff
@@ -62,7 +66,7 @@
                 <template
                   v-if="props.data.zoneIngress.advertisedSocketAddress.length > 0"
                 >
-                  <TextWithCopyButton
+                  <XCopyButton
```
[the remark this answers is not in the document]
Comment (schogges) [reply src/app/zone-ingresses/views/ZoneIngressDetailView.vue:69]: Nice! 👍 😄

File: src/app/subscriptions/data/index.ts [hunk 11]
```diff
@@ -1,32 +1,76 @@
...
-  subscriptions: T[]
-  connectedSubscription?: T
-  version?: T['version']
-}
-export type DiscoverySubscriptionCollection = SubscriptionCollection<DiscoverySubscription>
 
+export const Subscription = {
+  fromObject<T extends PartialSubscription>(item: T) {
+    return {
+      $raw: item,
```
Comment (schogges) [inline src/app/subscriptions/data/index.ts:29]: We usually name this key `config`, correct?

File: src/test-support/mocks/src/zone-ingresses/_/_overview.ts [hunk 12]
```diff
@@ -57,27 +62,41 @@ export default ({ fake, env }: EndpointDependencies): MockResponder => (req) =>
...
+                  const stats = Object.entries(fake.kuma.partitionInto(
+                    {
+                      cds: Number,
+                      eds: Number,
+                      lds: Number,
+                      rds: Number,
+                    }, xcks,
+                  ))
+                  return {
+                    lastUpdateTime: '2021-07-13T09:03:11.614941842Z',
```
Comment (schogges) [inline src/test-support/mocks/src/zone-ingresses/_/_overview.ts:77]: Should we use a random fake date in the past? (same for the other cases)

**Review by schogges — APPROVED** [review]
👍 
left some Qs 🙂

File: src/app/kuma/locales/en-us/index.yaml (hunk 2 above)
Comment (johncowen) [reply src/app/kuma/locales/en-us/index.yaml:103]: A while back we put i18n in place (we bought shovels), we just went crazy digging everywhere instead of finding the right place to dig first. Once we've found the right place, we can dig there, but we'll have to also fill up all the places where we dug in the wrong place i.e. the refactor. Not sure if its a good idea to continue with the analogy! 😆 

---

To give some history: I was really keen to get i18n in place, but at first **purely so we could do "machine to human" translations**. So for example we use it a lot for our tabs where we use the `route.name` as the i18n key for the text in the tab, it translates the "machine" `route.name` to "human" text (not necessarily of any particular language). A lot of this "machine" to "human" was converting HTTP API properties/values to "human" (e.g. `mtls` to `mTLS`). I think we did this right by having `http.api.property`, `http.api.value` keys to have centrally reference-able human texts for these things.

Thing is, once we added i18n, we then went ahead and did lots of localisation on things that weren't to do with "machine" to "human" translation, and we did the i18n strings using completely different conventions, so there is no consistency anywhere. In some places we didn't think about using the `http.api.property` section of our i18n strings. i.e. we dug lots of different holes all over the place, with very few in the right place. i.e. the convention (or lack of convention) we have is all wrong. We are going to have to go back and change it all, the more there is the more we'll have to change.

So basically exactly what you said "we have to find a good pattern and structure", a lot of it doesn't have that as yet, so once we find that pattern (the sooner the better) we'll have to go through and apply that pattern to all of the places where we don't use it.

Hope that helps! Analogies are sometimes a bad idea 😬 😆

File: src/app/subscriptions/data/index.ts (hunk 11 above)
Comment (johncowen) [reply src/app/subscriptions/data/index.ts:29]: Yeah I bumped into this during the work. A `subscription` already has a `config` property, so we can't use that name 😬 .

We wrote about this eventuality somewhere and said we could use `$raw` if we ever bumped into this (and in other places we already use `$raw`). We could change the places where we use `config` for this so that they all use `$raw` really so its all consistent.

Lemme see if I can find where we wrote about this

File: src/app/zone-ingresses/views/ZoneIngressDetailView.vue (hunk 10 above)
[the remark this answers is not in the document]
Comment (johncowen) [reply src/app/zone-ingresses/views/ZoneIngressDetailView.vue:69]: 👍 I made an issue for this work https://github.com/kumahq/kuma-gui/issues/3157, gonna resolve this convo

File: src/test-support/mocks/src/zone-ingresses/_/_overview.ts (hunk 12 above)
Comment (johncowen) [reply src/test-support/mocks/src/zone-ingresses/_/_overview.ts:77]: Yeah defo, I think (from vague recollection) the reason dates have been an outlier in the mocks is that faker doesn't offer a helper that formats dates like these golang formatted dates. Lemme take a quick look and see if thats correct. We should add one of our own formatters if so and use it everywhere.

Comment (johncowen) [reply src/test-support/mocks/src/zone-ingresses/_/_overview.ts:77]: Oh! I just looked so there's code for this in that `fake.kuma.connection` thing we were chatting over the other day. I can extract that so we can use it here also, will be much better. I'll PR that separately in a sec.

Thanks for the suggestion! 🎉

Comment (johncowen) [reply src/test-support/mocks/src/zone-ingresses/_/_overview.ts:77]: I made this https://github.com/kumahq/kuma-gui/pull/3158, gonna resolve this convo

Comment (schogges) [reply src/test-support/mocks/src/zone-ingresses/_/_overview.ts:77]: Perfect 👍 🙂

**Review by lahabana — APPROVED** [review]

File: src/app/subscriptions/data/index.ts (hunk 11 above)
Comment (schogges) [reply src/app/subscriptions/data/index.ts:29]: I see 🙂 I like consistency, so I'd vote for making use of `$raw` as this key is more likely not going to appear in the api payloads 🙂

File: src/app/kuma/locales/en-us/index.yaml (hunk 2 above)
Comment (schogges) [reply src/app/kuma/locales/en-us/index.yaml:103]: Alright, got it, thank you 🙂 👍
